### PR TITLE
Separate event Publisher from Subscriber.

### PR DESCRIPTION
**Pull request by doychin**
Signed-off-by: Doychin Bondzhev <doychin@dsoft-bg.com>

### CODE REVIEW

**Comment by otaviojava** [comment]
Hey, @doychin did you check the newest Java 9 API?
It has new support to flow.
http://download.java.net/java/jdk9/docs/api/java/util/concurrent/Flow.html

That might be a good opportunity to check it.

**Comment by aalmiray** [comment]
@otaviojava for now JSR377 will use JDK8 as source/target compatibility, thus we can't use any new APIs brought by JDK9. Also, Flow doesn't exactly solve the problem of firing events in a synchronous way, as the whole point of Flow is to allow a series of values to be sent to potential listeners on an asynchronous way.

This is the trap that https://github.com/square/otto feel into, the authors believe you can model an event bus using Observables (Rx) but you can only do if you only care about asynchronous notifications. Another problem with this approach is that event handlers must register to an specific event source which supplies values of a single type; this means a component that needs to handle events of type `EventOne` and `EventTwo` must register with two sources, instead of one.

You also have the problem of cold observables vs hot observables.

**Comment by doychin** [comment]
For me the problem with single interface is that it is not clear for the end user what he will do with this instance. I understand that the implementation will need to implement both interfaces but for end user it is much clearer when it has to consume events or fire events. It will inject both in case it is both consumer and generator of events. If it will generate events on another bus separating these will make it clear where code will be in the role of consumer and where it will be in the role of source of events. As you say in your other message a qualifier can be used for that purpose.

**Comment by doychin** [comment]
Of course if the naming is not correct we can think of better name for EventSubscriber interface. 
For example EventProducer.

So in the code you inject EventProducer and you subscribe/unsubscribe to it's events.

File: api/src/main/java/javax/application/event/EventSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * Base contract for classes that will provide event handlers using their own
+ * event bus.
+ */
+public interface EventSubscriber {
+    /**
+     * Adds an event handler.<p>
+     *
+     * @param handler an event handler. Must not be {@code null}.
+     */
+    void subscribe(Object handler);
```
Comment (jbuddha) [inline api/src/main/java/javax/application/event/EventSubscriber.java:28]: Can this parameter type be EventHandler type? Same comment for the unsubscribe method as well.

Comment (doychin) [reply api/src/main/java/javax/application/event/EventSubscriber.java:28]: No, it is not possible. EventHandler is annotation.  You use it to annotate the methods in handler object that will be executed when an event of specific type is triggered.

Comment (jbuddha) [reply api/src/main/java/javax/application/event/EventSubscriber.java:28]: Sorry, you are right, I got confused.

**Pull request closed without merging**